build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts

```diff
@@ -1416,6 +1416,108 @@ export const handlers: Handler[] = [
 				}
 			});
 
+			return result;
+		},
+	},
+	{
+		// Verify that non-private packages are configured correctly for API-Extractor by checking for the following:
+		// - dev dependency on `@microsoft/api-extractor`
+		// - `build:docs` script
+		// - `check:release-tags` script
+		name: "api-extractor-public-packages",
+		match,
+		handler: async (file) => {
+			let json: PackageJson;
+			try {
+				json = JSON.parse(readFile(file));
+			} catch (err) {
+				return "Error parsing JSON file: " + file;
+			}
+
+			const packageName = json.name;
+
+			if (json.private || packageName.startsWith("@fluid-private")) {
+				// If the package is private, we have nothing to validate.
+				return;
+			}
+
+			const errors: string[] = [];
+
+			// Ensure the package has a dev dependency on `@microsoft/api-extractor`
+			const devDependencies = Object.keys(json.devDependencies ?? {});
+			if (!devDependencies.includes("@microsoft/api-extractor")) {
+				errors.push(`Missing dev dependency: "@microsoft/api-extractor"`);
+			}
+
+			/**
+			 * Creates an error string noting the package.json is missing a script of the provided name.
+			 * @param missingScriptName - The name of the missing script.
+			 */
+			function createMissingScriptErrorString(missingScriptName): string {
+				return `Missing script: "${missingScriptName}"`;
+			}
+
+			const scripts = Object.keys(json.scripts ?? {});
+
+			// Ensure `build:docs` script exists
+			if (!scripts.includes("build:docs")) {
+				errors.push(createMissingScriptErrorString("build:docs"));
+			}
+
+			// Ensure `check:release-tags` script exists
+			if (!scripts.includes("check:release-tags")) {
+				errors.push(createMissingScriptErrorString("check:release-tags"));
+			}
+
+			// Ensure `ci:build:docs` script exists
+			if (!scripts.includes("ci:build:docs")) {
+				errors.push(createMissingScriptErrorString("ci:build:docs"));
+			}
+
+			// TODO: any other scripts to check? `apis`? `api-extractor:*`?
+
+			if (errors.length > 0) {
+				return [
+					`API-Extractor violations for public package "${json.name}":`,
+					...errors,
+				].join(`${newline}* `);
+			}
+		},
+		// Fixes the following issues:
+		// - Adds missing `build:docs` script
+		// - Adds missing `check:release-tags` script
+		// TODO: we don't currently have a nice mechanism for auto-installing missing deps from here,
+		// but we could in theory install a version that matches what's in the base package.json or something.
+		resolver: (file) => {
+			const result: { resolved: boolean; message?: string } = { resolved: true };
```

Comment:
Seems like updating this before we return it would be nice, especially to point out if any policy violations couldn't be fixed, like the new devDependency. Not sure if `resolved` should return true or false in that case, though.

Reply:
> we don't currently have a nice mechanism for auto-installing missing deps from here

Not sure it's a good idea, but if you exec `pnpm add -D dependency` in the package folder pnpm will use the correct version. We shell out to pnpm in other places so it's not too crazy. But I think it's also fine to not fix this automatically. In my experience, the more complex the handlers/resolvers get, the less useful they end up being.

Reply:
Good point. Yeah, presumably this wants to return false if the dev dependency is missing. Will update.

Reply:
For now, I left the installation step to be manual, but the code does now set `resolved` to false in these cases.